edb/edgeql/compiler/triggers.py

```diff
@@ -0,0 +1,150 @@
+            schema, source, typename=sn.QualName(
+                module='__derived__', name='__new__')
+        )
+        new_set = setgen.class_set(
+            source, path_id=new_path, ignore_rewrites=True, ctx=sctx)
+
+        old_set = None
+        if qltypes.TriggerKind.Insert not in kinds:
+            old_path = irast.PathId.from_type(
+                schema, source, typename=sn.QualName(
+                    module='__derived__', name='__old__')
+            )
+            old_set = setgen.class_set(
+                source, path_id=old_path, ignore_rewrites=True, ctx=sctx)
+            old_set.expr = irast.TriggerAnchor(typeref=old_set.typeref)
+            anchors['__old__'] = old_set
+        if qltypes.TriggerKind.Delete not in kinds:
+            anchors['__new__'] = new_set
+
+        for name, ir in anchors.items():
+            if scope == qltypes.TriggerScope.Each:
+                sctx.path_scope.attach_path(ir.path_id, context=None)
+                sctx.iterator_path_ids |= {ir.path_id}
+            sctx.anchors[name] = ir
+
+        trigger_set = dispatch.compile(
+            trigger.get_expr(schema).qlast, ctx=sctx)
+
+    typeref = typegen.type_to_typeref(source, env=ctx.env)
+    taffected = {
+        (typegen.type_to_typeref(t, env=ctx.env), ir) for t, ir in affected
+    }
+
+    return irast.Trigger(
+        expr=trigger_set,
+        kinds=kinds,
+        scope=scope,
+        source_type=typeref,
+        affected=taffected,
+        new_set=new_set,
+        old_set=old_set,
+    )
+
+
+def compile_triggers(
+    dml_stmts: Collection[irast.MutatingStmt],
+    *,
+    ctx: context.ContextLevel,
+) -> tuple[irast.Trigger, ...]:
+    schema = ctx.env.schema
+
+    trigger_map: dict[
+        s_triggers.Trigger,
+        set[tuple[s_objtypes.ObjectType, irast.MutatingStmt]],
+    ] = {}
+    for stmt in dml_stmts:
+        kind = TRIGGER_KINDS[type(stmt)]
+
+        stype = schemactx.concretify(
+            setgen.get_set_type(stmt.result, ctx=ctx), ctx=ctx)
+        assert isinstance(stype, s_objtypes.ObjectType)
+        # For updates and deletes, we need to look to see if any
+        # descendant types have triggers.
+        if isinstance(stmt, irast.InsertStmt):
+            stypes = {stype}
+        else:
+            stypes = schemactx.get_all_concrete(stype, ctx=ctx)
+
+        # Process all the types, starting with the base type
+        for subtype in sorted(stypes, key=lambda t: t != stype):
+            for trigger in subtype.get_relevant_triggers(kind, schema):
+                mro = (trigger, *trigger.get_ancestors(schema).objects(schema))
+                base = mro[-1]
+                tmap = trigger_map.setdefault(base, set())
+                # N.B: If the *base type* of the DML appears, that suffices,
+                # because it covers everything, and we don't need to duplicate.
+                if (stype, stmt) not in tmap:
+                    tmap.add((subtype, stmt))
```

Comment:
I think this could be "If *any of the previously handled triggers of ancestors* appears, ...

This would have effect when you have a trigger on `Child1` in hierarchy `Updated > Child1 > Child2`. 
You would first add the trigger on Child1, but when processing Child2, the trigger would be added again, even though it's effects have already been covered.

(I may not be understanding this whole chunk correctly)

Reply:
Yeah, it is a specific interaction with how the pgsql/dml.py side is implemented. I'm elaborating a bit more in the comment.